initial pr

johnmarinelli opened the pull request:
new relic dashboards and fullscreen api implemented

dscataglini commented on app/services/new_relic_api.rb:
```diff
@@ -0,0 +1,130 @@
+        db_response_time = (ar_response_time * ar_call_count) / hd_call_count
+
+        if db_response_time.infinite?
+          puts '============================= ^_^ LOL UR FUCKED ==================================================='
```
????

dscataglini replied:
This might work on a command line level. You should use Rails.logger.debug

johnmarinelli commented:
ya; I was going to actually email new relic because that *technically* shouldn't be happening.  according to their docs, I'm using the correct equation.  however now that we're using embedded charts, I don't think I'll even need this function

dscataglini commented on app/controllers/dashboards_controller.rb:
```diff
@@ -0,0 +1,87 @@
+  # GET /dashboards
+  # GET /dashboards.json
+  def index
+    @dashboards = Dashboard.all
```
Always paginate

johnmarinelli replied:
only paginate for :index, right?

dscataglini replied:
yes

dscataglini commented on app/controllers/dashboards_controller.rb:
```diff
@@ -0,0 +1,87 @@
+
+  private
+    # Use callbacks to share common setup or constraints between actions.
+  def set_dashboard
```
A different, clearer way to do this is not to use it as filter but as an accessor

    def current_dashboard
      @dashboard ||= Dashboard.find(params[:id])
    end

johnmarinelli replied:
good idea; changed

johnmarinelli replied:
actually idk if it should really be set to Dashboard.new since it's being called before :show, :edit, :update, and :destroy, all of which assume an existing dashboard.  if i call dashboards/#{SOME_NONEXISTENT_ID}/edit, then i'll just get an error page as normal.

dscataglini commented on app/controllers/dashboards_controller.rb:
```diff
@@ -0,0 +1,87 @@
+  # GET /dashboards/1
+  # GET /dashboards/1.json
+  def show
+    @airbrake_groups = @dashboard.airbrake_charts.map do |a|
```
A lot is happening in this method, all related to airbrake_groups.
It seems this might be a helper method, separate project, or a method on the Dashboard class
ie:

    def airbrake_groups_from_dashboard(dashboard = Dashboard.new)
        …

or

    class Airbrake < Struct.new(:dashboard)
      …

or 

    class Dashboard < ActiveRecord::Base
      def airbrake_groups
        …

johnmarinelli replied:
i see what you mean.
i'll move it into a helper method in the controller or just to an external object.
all it does is transform data to be displayed in the view, which is why i'm hesitant to put it in the model

dscataglini commented on app/controllers/dashboards_controller.rb:
```diff
@@ -0,0 +1,87 @@
+  # DELETE /dashboards/1
+  # DELETE /dashboards/1.json
+  def destroy
+    @dashboard.destroy
```
What if @dashboard is not found?

johnmarinelli replied:
redirect to index with notice, what do you think

johnmarinelli replied:
addressed above; if dashboard isn't found there is an error page

dscataglini commented on app/services/airbrake_api.rb:
```diff
@@ -0,0 +1,33 @@
+    groups = groups_api.get_groups project_id
+ end
+
+  class API < API::BaseAPI
```
You could be clearer on where this is coming from by saying ::API::BaseAPI

johnmarinelli replied:
agree, didn't even notice that

dscataglini commented on app/services/api.rb:
```diff
@@ -0,0 +1,26 @@
+
+    def create_get_request(url)
+      create_http url
+      @req = Net::HTTP::Get.new url
```
Does this take account of redirections? or failures?

johnmarinelli replied:
not yet; will do

dscataglini commented on app/services/airbrake_api.rb:
```diff
@@ -0,0 +1,33 @@
+    def get_groups(project_id)
+      url = URI.join @api_url, "#{project_id}/", 'groups'
+      url.query = URI.encode_www_form @query_params
+      create_get_request url
```
create_get_request populates a @http and @req. I get the @req but the @http comes out of nowhere.
I'd have been happier with httparty or openuri used everywhere

johnmarinelli replied:
i'll move all the net::http stuff to openuri

dscataglini replied:
Actually HTTPParty might be better, check

dscataglini commented on app/services/airbrake_api.rb:
```diff
@@ -0,0 +1,33 @@
+      url.query = URI.encode_www_form @query_params
+      create_get_request url
+      res = @http.request @req
+      groups = JSON.parse res.body
```
Response might be invalid JSON

johnmarinelli replied:
i think this ties into the redirect/error handling mentioned above; will do

dscataglini commented on app/services/airbrake_api.rb:
```diff
@@ -0,0 +1,33 @@
+      url = URI.join @api_url, "#{project_id}/", 'groups'
+      url.query = URI.encode_www_form @query_params
+      create_get_request url
+      res = @http.request @req
```
This might fail

dscataglini commented on app/services/new_relic_api.rb:
```diff
@@ -0,0 +1,117 @@
+        }
+      }
+    rescue StandardError => e
+      puts e
```
puts to what?

johnmarinelli replied:
going to trim the fat/remove all the puts

dscataglini commented on app/services/new_relic_api.rb:
```diff
@@ -0,0 +1,117 @@
+        db_response_time = (ar_response_time * ar_call_count) / hd_call_count
+
+        if db_response_time.infinite?
+          puts i
```
use Rails.logger instead of puts or insert into a database the error (although we could integrate with Airbrake and just post all the info to Airbrake :)

dscataglini commented on app/services/new_relic_api.rb:
```diff
@@ -0,0 +1,117 @@
+        
+        db_response_time = (ar_response_time * ar_call_count) / hd_call_count
+
+        if db_response_time.infinite?
```
All of this block can be captured in a different way

johnmarinelli replied:
not using this block anymore; going to remove

dscataglini commented on app/services/new_relic_api.rb:
```diff
@@ -0,0 +1,117 @@
+    dt.utc.strftime("%Y-%m-%dT%k:%M:%S%:z")
+  end
+
+  class API < API::BaseAPI
```
It seems to me, these embedded API are unnecessary. If you clean up the Base API none of these subclasses are necessary

dscataglini commented on app/services/new_relic_api.rb:
```diff
@@ -0,0 +1,117 @@
+    def get_app_metrics(app_id, name, values)
+      url = URI.join @api_url, "#{app_id}/", 'metrics/', 'data.json/' 
+      create_post_request url, { 'names[]' => name, 'values[]' => values }
+      res = @http.request @req
```
Same objections as above

dscataglini commented on app/services/new_relic_api.rb:
```diff
@@ -0,0 +1,117 @@
+
+        transformed_data
+      else
+        throw metrics['error']['title']
```
Are you sure you didn't mean raise?

dscataglini commented on app/services/new_relic_api.rb:
```diff
@@ -0,0 +1,117 @@
+          :db_response_time => db_response_time
+        }
+      }
+    rescue StandardError => e
```
This rescue will never rescue the throws that you're sending https://coderwall.com/p/lhkkug/don-t-confuse-ruby-s-throw-statement-with-raise

johnmarinelli commented:
1) moved the Net::HTTP stuff to HTTParty
2) took out that massive new relic block
3) took out the embedded API classes and the module
4) moved the dashboards#show internals to app/controllers/concerns/dashboards.rb
